deploy/docker/cp-tools/research/cellprofiler-web-api/app/src/hcs_manager.py

```diff
@@ -153,6 +164,186 @@ def _collect_parent_pipeline_inputs(self, parent_pipeline: HcsPipeline, projecti
+                clips.append(np.array(merged_image))
+
+        clip_path = CELLPROFILER_API_COMMON_RESULTS_DIR + self.get_clip_file_dir(path)
+        self._mkdir(clip_path)
+        clip_name = self.get_clip_file_name(by_field, cell, clip_format, sequence_id)
+        clip_full_path = clip_path + clip_name
+        slide = ImageSequenceClip(clips, fps)
+        slide.write_videofile(clip_full_path, codec=codec)
+        t1 = time.time()
+        return clip_full_path, round((t1 - t), 2)
+
+    @staticmethod
+    def get_clip_file_dir(path):
+        hcs_file_base_name = os.path.basename(path)
+        hcs_file_name = os.path.splitext(hcs_file_base_name)[0]
+        return '/video/{}/'.format(hcs_file_name)
+
+    @staticmethod
+    def get_clip_file_name(by_field, cell, clip_format, sequence_id):
+        cell_prefix = ('field{}' if by_field else '/well{}').format(str(cell))
+        sequence_prefix = ('' if sequence_id is None else 'seq{}').format(sequence_id)
+        return cell_prefix + sequence_prefix + clip_format
+
+    @staticmethod
+    def get_pages(channel_ids, channels, timepoints, cell):
+        pages = []
+        num = 0
+        for time_point in range(timepoints):
+            for channel_id in range(channels):
+                if channel_id in channel_ids:
+                    pages.append(num + channels * timepoints * cell)
+                num = num + 1
+        return pages
+
+    @staticmethod
+    def extract_xml_schema(xml_info_root):
+        full_schema = xml_info_root.tag
+        return full_schema[:full_schema.rindex('}') + 1]
+
+    @staticmethod
+    def parse_hcs(path, sequence_id):
+        with open(path) as json_file:
+            data = json.load(json_file)
+            preview_dir = data['previewDir']
+            time_series_details = data['time_series_details']
+            sequences = dict()
+            if sequence_id is not None:
+                sequences[sequence_id] = time_series_details[sequence_id]
+            else:
+                sequences = time_series_details
+            return preview_dir, sequences
+
+    def get_all_channels(self, path):
+        tree = ET.parse(path)
+        hcs_xml_info_root = tree.getroot()
+        hcs_schema_prefix = self.extract_xml_schema(hcs_xml_info_root)
+        maps = hcs_xml_info_root.find(hcs_schema_prefix + 'Maps').findall(hcs_schema_prefix + 'Map')
+        channels = []
+        for m in maps:
+            entries = m.findall(hcs_schema_prefix + 'Entry')
+            for e in entries:
+                channel_name = e.find(hcs_schema_prefix + 'ChannelName')
+                if channel_name is not None:
+                    channel = dict()
+                    channel['id'] = int(e.attrib.get('ChannelID')) - 1
+                    channel['name'] = channel_name.text
+                    channels.append(channel)
+        return channels
+
+    @staticmethod
+    def extract_local_path(path, cloud_scheme=HCS_CLOUD_FILES_SCHEMA):
```

Comment:
It looks like this logic already exists in `hcs_module_factory.py` method `prepare_input_path`. Can we reuse it somehow?